1008 ideal integration online payment provider

yksflip opened the pull request:
This PR adds the online payment provider _mollie_ as a plugin. User can add credit to their ordergroups balance via various payment methods. wvengen implemented this already in the foodsoft-adams fork. I rewrote some parts to make it work with the newest mollie api version and added some bits to improve the usability.

Changes relevant to the foodsoft core:
* added new attributes to the financial_transaction model to keep track of the state of a transaction
* financial_transaction.value may be `nil` now, this indicates an incomplete transaction

Changes on the plugin:
* payment_method is no longer choosen in the foodsoft form but on the mollie checkout
* information regarding the transaction fee is displayed 
* settings can be configured in the admin screen
* the transaction_type for mollie transactions can be configured in the administration settings

still open
* ~~[ ] move [accept_return_to](https://github.com/foodcoops/foodsoft/pull/1031/files#diff-5856be838b4186db9c017a6d2a3211e100851ae2c5418e4f52f29ddeec70e353R111) to ApplicationController?~~
* [x] open currency refactoring issue, implement FoodsoftMollie.currency
* [x] write financial transaction documentation
* [x] reverting a pending financial transaction?
* [x] handle amount < min ?
* [x] squash commits

wvengen reviewed (commented):
Wow, great to see this coming back, now properly as a core Foodsoft plugin! :)

It's been a long time, probably some questions I'm asking are about pieces you've just copied from the original code. Nevertheless, it's good to remain critical (also on my past self).

There are two possibly important assumptions that change:
- A FinancialTransaction's `amount` may be `nil`
- A FinancialTransaction is mutable

This is something important to think through. Solid accounting has no changing history, but in this case we make an exception. I think this is justified, but I'm not 100% sure - both in Foodsoft core (not all parts of the code are tested fully), and by plugins/integrations people have written. This even affects the API (financial transations), where software using it may not expect `nil` for amount.
I think before this PR is merged, someone needs to have a good look at this. Perhaps you've done that already.

Do we need to make a new API version for this, hide the unfinished transactions by default, or accept a possibly incompatible change (the API is relatively young, so maybe ok)?
One way to get an answer to this is to ask this question on the forum and/or Github.

I also asked some questions to which I could have found the answer myself. Hope that's ok, and you're patient enough to go look, think about it, and respond.

Some general things to improve could be (partly already mentioned in separate comments):
- Use strong parameters, they're standard now, and with finance-related code it's good to be safe.
- I think it's good to document what we expect from financial transactions, just some background in the main Foodsoft (I guess in-tree) developer docs, about the assumptions. Maybe it's not really a question for you. I think it would have best been done already with the introduction of financial transaction types and such. But this PR touches something quite important, so I think writing down something about financial transactions may help even getting your own mental model right.

Please feel free to disagree too, I hope we can have a constructive conversation :)
Thanks for all your work in making this ready for prime-time!

wvengen commented:
p.s. if the `amount` can be `nil` in the API, the API docs (swagger) may need updating.

yksflip commented on db/migrate/20230915093041_add_payment_to_financial_transaction.rb:
```diff
@@ -0,0 +1,26 @@
+    add_column :financial_transactions, :payment_state, :string
+    add_column :financial_transactions, :payment_fee, :decimal, :precision => 8, :scale => 3
+    add_column :financial_transactions, :payment_acct_number, :string
+    add_column :financial_transactions, :payment_acct_name, :string
```
Yes you're right. I think for the mollie usecase we actually don't need to save this information, as it's not used anywhere. And if we wan't to it could still be fetched from the api...

yksflip commented on plugins/mollie/README.md:
```diff
@@ -0,0 +1,37 @@
+      ideal: 1.20
+      banktransfer: 0.30
+      creditcard: 3.39% + 0.05
+      paypal: 0.18 + 0.35 + 3.4%
```
I think this is quite tricky. There is a way to get fee's from the mollie api, but e.g. paypal charges additional fees ... But maybe this is only the exception. 

Maybe this would be a sane config interface: 
* [] Charge ordergroup mollie fee's 
  * [] defaults from mollie-api
  * Input custom fee for  $payment_method

yksflip replied:
I thought a override might be useful if the actual costs differ for some reason (I've read smth about paypal having extra fee's, but I'm not sure). But yeah maybe overkill for the moment.

Via the [/methods endpoint](https://docs.mollie.com/reference/v2/methods-api/get-method#example) it is possible to get mollies fees. And yess it's a bit more complex :o
Though I haven't found a way to get a computed fee from mollie payment result.

Having the option to topup the fee amount would be great. I think I'll add the method and issuer selection again, so the fee can be computed beforehand.

yksflip commented on plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb:
```diff
@@ -0,0 +1,160 @@
+    return unless params[:return_to].starts_with?(root_path) || params[:return_to].starts_with?(root_url)
+
+    session[:return_to] = params[:return_to]
+  end
```
tbh this is a copy pasta of your code :o
but I'll open a refactoring issue :)

yksflip replied:
I choose to delete the return_to functions completely, afaik we don't need it atm.

yksflip commented on plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb:
```diff
@@ -0,0 +1,160 @@
+      },
+      description: "#{@transaction.id}, #{@ordergroup.id}, #{FoodsoftConfig[:name]}",
+      redirectUrl: result_payments_mollie_url(id: @transaction.id),
+      webhookUrl: (request.local? ? check_payments_mollie_url : 'https://example.com'), # Mollie doesn't accept localhost here
```
whoops, it was actually the opposite,  hope its more clear now?

wvengen replied:
Ah, that makes more sense!

yksflip commented on plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb:
```diff
@@ -0,0 +1,160 @@
+    payment = Mollie::Payment.create(
+      amount: {
+        value: format('%.2f', amount),
+        currency: 'EUR' # @todo payment_currency
```
I guess this would be a big refactoring ... I'd open a issue and reference it in the according places. I could also make it call smth like `FoodsoftMollie.currency`, than later only one place needs to be changed

wvengen replied:
What about making this a Mollie configuration option (with a default to `EUR`), explaining that this needs to match Foodsoft's currency?
Available currencies at: https://docs.mollie.com/payments/multicurrency

yksflip replied:
I had to realise that mollies method endpoint will always return the fee amount in EUR. I think for now the most straight forward thing is to make 'EUR' a constraint for the fee feature (I'd guess that's the most common case). If we find a way to retrieve the actual fee calculated by mollie we can move on to that. (Still calculating the right fee to increase the amount for different currencies before paying would be difficult)

yksflip commented on plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb:
```diff
@@ -0,0 +1,160 @@
+    Mollie::Client.configure do |config|
+      config.api_key = FoodsoftConfig[:mollie]['api_key']
+    end
+  end
```
I have no idea, I'm really not deep enough into RoR yet. I don't even know how to figure that out ... Can you give me another hint? :)

wvengen replied:
See `set_currency` and `set_time_zone` in the ApplicationController.

An around hook would probably be best, setting it early in the request/response process, and clearing it afterwards.
Another option would be to set it like here, but then it needs to be set _always_ (maybe even when the plugin is disabled?), to be sure it doesn't look. An around hook is safer.

yksflip replied:
I don't know why, but I've got 401 unauthorized errors when trying to this as a around hook. I've decided to just pass the api-key in the function calls directly (it's only twice)

yksflip commented on plugins/mollie/app/overrides/admin/configs/_tab_others/add_config.html.haml.deface:
```diff
@@ -0,0 +1,12 @@
+= config_input form, :use_mollie, as: :boolean
+= form.fields_for :mollie do |fields|
+  = config_input fields, :api_key, input_html: {class: 'input-xlarge'}
+  = config_input fields, :financial_transaction_type, :as => :select, :collection => FinancialTransactionType.order(:name).map { |t| [ t.name, t.id ] }
```
yes, maybe that could be the case. This is the way vokoosterwold would like to use it, but there might be other use cases. I think once they appear we should try to add them. 
I'll try to learn more about the usage of ft types

wvengen commented on doc/API.md:
```diff
@@ -15,14 +15,14 @@ It can be generated running `RAILS_ENV=test rails rswag`.
 You'll need to give access to the application first. This can be done by going to
 _Administration_ > _Configuration_ > _Apps_ in Foodsoft. Select _New Application_,
-enter any name, put `http://petstore.swagger.io/oauth2-redirect.html` in _Redirect URI_
+enter any name, put `http://localhost:3000/api-docs/oauth2-redirect.html` in _Redirect URI_
```
Unrelated change, though useful :)

yksflip replied:
ahh couldn't resist, I'll put it in a separate commit.

wvengen reviewed (commented):
This is shaping up nicely!

yksflip commented:
> This is shaping up nicely!

thx! your comment are very helpful :)

yksflip commented:
Hey I've been working a bit more an introduced some changes:

  * transactions with nil value are no longer revertable
  * display payment_state in transaction views
  * strong params on api controller
  * api-key is now used in function call (around filter didn't work)
  * introduce currency config
    Mollie returns fee amounts always in EUR, no matter what currency is
    selected. Because of that there is a constraint to charge fees for
    EUR only.
  * introduce default amount config
  * some code refactoring

Hopefully I can finish up on this until end of this week :)

yksflip commented on app/helpers/application_helper.rb:
```diff
@@ -24,6 +24,16 @@ def format_currency(amount)
+    if t.amount
+      format_currency t.amount
+    elsif t.payment_amount && t.payment_state
+      content_tag :span, "#{number_to_currency t.payment_amount}) (#{t.payment_state})", class: :disabled_amount
```
I thought it would be the best to just display the payment state as a seperate column. It felt weird to use the amount column for this information.

yksflip commented on plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb:
```diff
@@ -0,0 +1,160 @@
+    amount = params[:amount].to_f
+    amount = [params[:min].to_f, amount].max if params[:min]
+
+    redirect_to new_payments_mollie_path(session[:mollie_params]), alert: t('.invalid_amount') and return if amount <= 0
```
I'd go with silently assuming the _min_. Still users have to confirm the amount again on Mollie page.

wvengen commented:
Please let me know when it is ready for review (comment or the re-request review button), as it's been a while. Would be great to have this included in the next release.

yksflip commented:
hey @wvengen, thanks for the reminder. I know it's been a while ... I tried to write some documentation, though I don't feel very confident with it. I still don't understand very much of the banking feature. I see this docs as a starting point...
I'd happy to hear your feedback and would be great to see this finally merged :)

wvengen commented on app/controllers/api/v1/financial_transactions_controller.rb:
```diff
@@ -4,7 +4,7 @@ class Api::V1::FinancialTransactionsController < Api::V1::BaseController
 
   def index
-    render_collection search_scope
+    render_collection params.permit(:include_incomplete) ? search_scope : search_scope.where.not(amount: nil)
```
What about moving the `include_complete` logic to `search_scope` in this class?

wvengen commented on app/controllers/api/v1/user/financial_transactions_controller.rb:
```diff
@@ -7,7 +7,7 @@ class Api::V1::User::FinancialTransactionsController < Api::V1::BaseController
 
   def index
-    render_collection search_scope
+    render_collection params.permit(:include_incomplete) ? search_scope : search_scope.where.not(amount: nil)
```
idem

wvengen commented on app/helpers/application_helper.rb:
```diff
@@ -20,7 +20,7 @@ def format_datetime_timespec(time, format)
   def format_currency(amount)
-    class_name = amount < 0 ? 'negative_amout' : 'positive_amount'
+    class_name = amount < 0 ? 'negative_amout' : 'positive_amount' if amount.is_a? Numeric
     content_tag :span, number_to_currency(amount), class: class_name
```
I think this works, but is a little strange: if amount is not numeric, then `class_name` is actually undefined.

yksflip replied:
should we set class_name to a default value then?

wvengen replied:
Is it about `amount` being `nil`? In that case, we can just omit the `span` entirely and return `nil` directly.

wvengen commented on app/views/finance/financial_transactions/_transactions.html.haml:
```diff
@@ -20,6 +20,11 @@
+      %th= heading_helper FinancialTransaction, :payment_method
+      %th= heading_helper FinancialTransaction, :payment_amount
+      %th= heading_helper FinancialTransaction, :payment_fee
+      %th= heading_helper FinancialTransaction, :payment_state
```
Is the table still readable with a number of added columns (with possibly long names)?
This is especially relevant for foodcoops not using a payment plugin.

wvengen replied:
If we don't need a config option, let's avoid it (as it's complex enough as it is). So if they're a straightforward way to derive this, yes. Looking at payment_plugin config seems sensible here, yes.

Could there be a situation where this information is still relevant but the payment plugin is disabled? E.g. a payment plugin was enabled, but now disabled? I think we can just hide it, because any payment-plugin-related things will fail anyway. But just checking. E.g. at some places we check if there's more than one financial transaction type, and only show certain things if there is (to avoid UI complexity when a feature is not used).

wvengen commented on spec/swagger_helper.rb:
```diff
@@ -186,7 +186,8 @@
                     format: :float,
-                    description: 'amount credited (negative for a debit transaction)'
+                    nullable: true,
+                    description: 'amount credited. Negative for a debit transaction, null for a incomplete transaction.'
```
typo: "a incomplete" -> "an incomplete"

wvengen commented on app/models/financial_transaction.rb:
```diff
@@ -73,4 +75,12 @@ def created_at
+
+  def update_ordergroup_balance
+    # @todo Make sure this transaction and the ordergroup update is in one database transaction.
+    #   It may be possible to use an around filter if needed.
```
Just checking in if this todo is something that needs to be checked.

wvengen commented on app/models/financial_transaction.rb:
```diff
@@ -38,12 +43,9 @@ def self.ransackable_associations(_auth_object = nil)
-
   def revert!(user)
+    raise 'Pending Transaction cannot be reverted' if amount.nil?
+
```
Not sure how good the testing of reverting transactions is, ideally we'd test this.
But as we also want to get this PR merged, feel free to skip if this feature has no testing at all.

wvengen reviewed (commented):
Thank you for working on this PR!
Some small comments, I think it would be almost ready for merging.

Super you added some documentation. I find the financial transaction types and such also not that easy to comprehend. I think it's mostly the Austrian foodcoops using this, so they can best provide input here. See also [this forum thread](https://forum.foodcoops.net/t/transactionclasses-and-types/374), here is an issue [showing a use-case](https://github.com/foodcoops/foodsoft/issues/781), and [this issue](https://github.com/foodcoops/foodsoft/issues/170) has also some background.

yksflip commented:
thanks for your feedback!
I've added some basic model tests for financial_transactions, refactored the search_scope, and implemented the around_save.
I also just had a chat with leo and he mentioned the user documentation on https://docs.foodcoops.net/de/documentation/admin/finances

wvengen commented on spec/models/financial_transaction_spec.rb:
```diff
@@ -0,0 +1,31 @@
+  before do
+    ft
+    ordergroup
+  end
```
you can use `let!` here if you want to make sure these things are created

wvengen commented on spec/models/financial_transaction_spec.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+
+  it 'fails when trying to revert a pending transaction (amount=nil)' do
+    ft.update_attribute(:amount, nil)
```
I'd suggest to create a financial transaction trait in the factory that makes it more expressive, and will allow it to function when we'd change behaviour (e.g. have an explicit _pending_ attribute or so).

```ruby
context "with a pending transaction" do
  let!(:ft) { create(:financial_transaction, :pending, ordergroup: ordergroup, amount: 20) }
  it "fails on revert" do
      expect { ft.revert!(ft.user) }.to raise_error(RuntimeError, 'Pending Transaction cannot be reverted')
  end
end
```

wvengen reviewed (commented):
Thanks for the changes! We're definitely getting there :rocket:
Great you found the documentation, I didn't know about this, thanks for sharing.

yksflip commented:
how about this? https://github.com/foodcoops/foodsoft/pull/1031/files#diff-f5c62182c669e1defeb29be1ca2beac61b02af50970b7b8b823708aa855221f7R23

wvengen commented on plugins/mollie/README.md:
````diff
@@ -0,0 +1,33 @@
+    currency: EUR # should match the foodcoop's currency
+```
+
+The transaction fee is sub will be added on each payment when set; it is not set by default,
````
What does "is sub" mean? (could it be that it used to be "is substracted from ...").
Also, "it is not set by default" -> "it is disabled by default" sounds clearer.

wvengen commented on plugins/mollie/config/locales/nl.yml:
```diff
@@ -0,0 +1,44 @@
+        fee: Selecteer de juiste transactiekosten
+      controller:
+        result:
+          notice: Er is €%{amount} bijgeschreven op uw account.
```
Please use the standard way to format currencies, as the currency may not be €.
Ok, later I read that Mollie (currently) assumes EUR (is that still the case?), so maybe not important. But the english language does not have the currency, so good to bring in harmony.

yksflip replied:
oh yes, thats now taken from the payment_currency field, just forgot to adapt the nl translation. 
The support for automatic fee calculation is currently only supported for EUR, because of api limitation

wvengen reviewed (commented):
Super! Some small notes, otherwise good to go, I think :rocket:

wvengen commented on app/views/finance/financial_transactions/_transactions.html.haml:
```diff
@@ -20,6 +20,12 @@
       - if FinancialTransactionType.has_multiple_types
         %th= heading_helper FinancialTransaction, :financial_transaction_type
       %th= sort_link_helper heading_helper(FinancialTransaction, :note), "note"
+      - if @financial_transactions.with_payment_plugin.any?
```
Wait ... somehow I must have looked at an old commit, or misread it. You're already doing this! :)

yksflip commented on app/views/finance/index.html.haml:
```diff
@@ -29,14 +29,18 @@
             %td= ft.note
-            %td.numeric= format_currency ft.amount
+            - if ft.payment_state
+              %td= ft.payment_state
```
yeah that wasn't very clever :D

yksflip commented on app/views/home/index.html.haml:
```diff
@@ -63,13 +63,16 @@
           - if FinancialTransactionType.has_multiple_types
             %th= heading_helper FinancialTransaction, :financial_transaction_type
           %th= heading_helper FinancialTransaction, :note
+          - if current_user.ordergroup.financial_transactions.with_payment_plugin.any?
```
yes, I now reuse the already loaded financial transactions

yksflip commented on app/views/finance/financial_transactions/_transactions.html.haml:
```diff
@@ -43,13 +49,19 @@
+          %td= t.payment_method
+          %td= format_currency t.payment_amount
+          %td= format_currency t.payment_fee
+          %td= t.payment_state
```
yes, I introduced a helper variable and use the same for th and td now

yksflip commented:
@wvengen let's merge? :)

wvengen reviewed (approved):
Yes! Many many thanks for completing this (and for keeping up with my feedback, requesting changes several times).

Feel free to press the merge button! (And I would suggest rebase merge in this case - and I would reword the commits to add "PR #1031" but maybe that's overthinking it.)

wvengen commented:
Ah, and for a future PR, it's good to mention the issue number in the PR _description_ (e.g. in this case "Implements #1008"), so that Github recognizes this and provides cross-references. The "1008" in the PR title didn't do this.

yksflip commented:
Thank you for also keeping up and your helpful advise :)
